### PR TITLE
secure-boot: suite passing in qemu

### PR DESCRIPTION
**Pull request by macpijan**

### CODE REVIEW

File: lib/secure-boot-lib.robot [hunk 1]
```diff
@@ -4,6 +4,21 @@ Documentation       Collection of keywords related to UEFI Secure Boot
...
+${GOOD_KEYS_SHA256}=        13de737ba50c8d14a88aaf5314a938fb6826e18ba8f337470ee490b17dd6bea8
+${NOT_SIGNED_URL}=          https://cloud.3mdeb.com/index.php/s/zmJXxGG4piGB2Me/download/not_signed.img
+${NOT_SIGNED_NAME}=         not_signed.img
+${NOT_SIGNED_SHA256}=       15dc0a250b73c3132b1d7c5f8e81f00cc34d899c3ddecbb838a8cd0b66c4f608
+${BAD_KEYS_URL}=            https://cloud.3mdeb.com/index.php/s/BJPbSqRH6NdbRym/download/bad_keys.img
+${BAD_KEYS_NAME}=           bad_keys.img
+${BAD_KEYS_SHA256}=         6da92bd97d4b4ca645fa98dcdfdc0c6876959e5b815a36f1f7759bc5463e7b19
+${BAD_FORMAT_URL}=          https://cloud.3mdeb.com/index.php/s/AsBnATiHTZQ6jae/download/bad_format.img
+${BAD_FORMAT_NAME}=         bad_format.img
+${BAD_FORMAT_SHA256}=       59d17bc120dfd0f2e6948a2bfdbdf5fb06eddcb44f9a053a8e7b8f677e21858c
```
Comment (Psotas) [inline lib/secure-boot-lib.robot:19]: Wouldn't it be better/easier to put all .cer files on one image/device? Is there any reason we don't do it this way?

Comment (Psotas) [reply lib/secure-boot-lib.robot:19]: What is the idea for platforms without PiKVM?
If everything were on one image, it is possible to use a USB drive (some tests require the drive to be plugged in anyway).
Another option is to use the installed Linux to upload the necessary files to the USB.

**Comment by Psotas** [comment]
@macpijan 
Please review my commit 7157a603ab31297546de1aebf71c759554465958

File: lib/secure-boot-lib.robot [hunk 2]
```diff
@@ -164,28 +164,25 @@ Enter Volume In File Explorer
...
+            ${new_entry}=    Read From Terminal
+            ${status}=    Run Keyword And Return Status
+            ...    Should Contain    ${new_entry}    ${target_volume}
+            IF    ${status} == ${TRUE}    BREAK
+            IF    ${in} == 19    Fail    Volume not found
+            Press Key N Times    1    ${ARROW_DOWN}
+        END
     END
-    # Press Key N Times    1    ${ENTER}
+    Press Key N Times    1    ${ENTER}
```
Comment (macpijan) [inline lib/secure-boot-lib.robot:185]: @Psotas This cannot be here as would result in not needed ENTER when the index != -1

File: lib/secure-boot-lib.robot (hunk 1 above)
Comment (macpijan) [reply lib/secure-boot-lib.robot:19]: > Wouldn't it be better/easier to put all .cer files on one image/device? Is there any reason we don't do it this way?

That was the existing state of this MR, I have just reworked it. Feel free to improve that creating one image. You can ask @filipleple how he created these images. It could be easier in fact to have them on  a single disk.

I have assigned them labels with: ` sudo mlabel -i .<IMAGE>`

Comment (macpijan) [reply lib/secure-boot-lib.robot:19]: > What is the idea for platforms without PiKVM?

No idea yet, single disk could be easier here indeed.

File: lib/secure-boot-lib.robot (hunk 2 above)
Comment (Psotas) [reply lib/secure-boot-lib.robot]: Fixed: https://github.com/Dasharo/open-source-firmware-validation/pull/118/commits/64cf0c128b140fd1d6ab1a2a04c30dbf4445546a

File: lib/secure-boot-lib.robot (hunk 1 above)
Comment (Psotas) [reply lib/secure-boot-lib.robot:19]: I think we can leave it for later, we already have an issue for it.
https://github.com/Dasharo/open-source-firmware-validation/issues/68

Comment (pietrushnic) [inline lib/secure-boot-lib.robot:19]: @macpijan I'm not huge fan of relying on cloud.3mdeb.com to be honest it would be much better to rely on 3mdeb.com subdirectories or maybe some dedicated repository. From what I see footprint of cloud.3mdeb.com grow and I don't know the reason why. We already faced mass 404 when links were broken. I know we rely our business of DES on that, but in long run I think this is wrong strategy. Why can't we use dl.3mdeb.com ?

Comment (macpijan) [reply lib/secure-boot-lib.robot:19]: > Why can't we use dl.3mdeb.com ?

It has limited access policy, and stores some critical data from other websites, so mistakes may be costly.

Perhaps the repo for blobs such as these would make sense.

Comment (pietrushnic) [reply lib/secure-boot-lib.robot:19]: @macpijan we can use other server or repo as you suggest.

Comment (macpijan) [reply lib/secure-boot-lib.robot:19]: We have more places like this. I have created an issue for improving that: https://github.com/Dasharo/open-source-firmware-validation/issues/121
In this MR we will focus on the SB functionality.

File: lib/usb-hid-msc-lib.robot [hunk 3]
```diff
@@ -16,7 +16,7 @@ Download ISO And Mount As USB
         Add USB To Qemu    img_name=${DL_CACHE_DIR}/${img_name}
     ELSE
         IF    "${DUT_CONNECTION_METHOD}" == "pikvm"
-            Upload Image To PiKVM    ${PIKVM_IP}    ${img_url}    ${img_name}
+            Upload Image To PiKVM    ${PIKVM_IP}    ${img_name}    ${img_url}
```
Comment (Psotas) [inline lib/usb-hid-msc-lib.robot:19]: `Upload Image PiKVM    ${pikvm_ip}    ${img_url}    ${img_name}    ${login}    ${password}` is the correct sequence

https://github.com/3mdeb/pikvm-rest-api/commit/529d434c1ac884f3cc5fa2e1a451829e95a83a0d
```
Upload Image To PiKVM
(...)
    [Arguments]    ${pikvm_ip}    ${img_url}    ${img_name}    ${login}=admin    ${password}=admin
    Upload Image PiKVM    ${pikvm_ip}    ${img_url}    ${img_name}    ${login}    ${password}
```
```
@keyword("Upload Image PiKVM")
def UploadImagePiKVM(
    pikvm_ip=str, img_url=str, img_name=str, login="admin", password="admin"
):
```

**Comment by macpijan** [comment]
@Psotas 

> @macpijan https://github.com/Dasharo/open-source-firmware-validation/commit/a0702d66935e491f3c4071b6d896cc8ab8bbdad8 this commit change anything?

You can check by yourself:

```
 git sh a0702d6
commit a0702d66935e491f3c4071b6d896cc8ab8bbdad8 (origin/qemu-secure-boot-rebase-pt2, qemu-secure-boot-rebase-pt2)
Author: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>
Date:   Mon Nov 20 09:56:05 2023 +0100

    pikvm-rest-api: update
    
    Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

diff --git a/pikvm-rest-api b/pikvm-rest-api
index 2de063f810ac..529d434c1ac8 160000
--- a/pikvm-rest-api
+++ b/pikvm-rest-api
@@ -1 +1 @@
-Subproject commit 2de063f810aced5fa90546481e5c68e93e2aed3b
+Subproject commit 529d434c1ac884f3cc5fa2e1a451829e95a83a0d
```

File: lib/usb-hid-msc-lib.robot (hunk 3 above)
Comment (macpijan) [reply lib/usb-hid-msc-lib.robot:19]: @Psotas Are you requesting some change here? If yes, please just push the commit and link here.

Comment (macpijan) [reply lib/usb-hid-msc-lib.robot]: @Psotas I dropped this commit from this MR.

**Comment by Psotas** [comment]
>You can check by yourself:
 git sh a0702d6

Thanks, now it's clear to me how github shows it.

File: lib/secure-boot-lib.robot [hunk 4]
```diff
@@ -175,113 +144,77 @@ Check Secure Boot In Windows
...
+    Enter Submenu From Snapshot    ${enroll_sig_menu}    Enroll Signature Using File
+
+Enter Volume In File Explorer
+    [Documentation]    Enter the given volume
+    [Arguments]    ${target_volume}
+    # 1. Read out the whole File Explorer menu
+    ${volumes}=    Get Submenu Construction    opt_only=${TRUE}
+    Log    ${volumes}
+    # 2. See if our label is within these entries
+    ${index}=    Get Index Of Matching Option In Menu    ${volumes}    ${target_volume}    ${TRUE}
```
Comment (Psotas) [inline lib/secure-boot-lib.robot:167]: regarding the comment
https://github.com/Dasharo/open-source-firmware-validation/commit/7157a603ab31297546de1aebf71c759554465958#r133696812
Probably if we use keywrod here
`Get Index Of Matching Option In Menu` then you will be able to undo the changes in the keyword itself.

**Comment by Psotas** [comment]
https://github.com/Dasharo/open-source-firmware-validation/blob/2b4780c0b126736dd48ce12aa9eee6d39966d888/keywords.robot#L70-L83
Replace:
`Set Global Variable    ${DUT_CONNECTION_METHOD}    SSH`
by:
`Set Test Variable    ${DUT_CONNECTION_METHOD}    SSH`
I think this will work properly.
If this is not there, the next tests will skip on `Download ISO And Mount As USB`  when check on `IF "${DUT_CONNECTION_METHOD}" == "pikvm"`

Logs:
[secure-boot_log_skip_global.zip](https://github.com/Dasharo/open-source-firmware-validation/files/13561116/secure-boot_log_skip_global.zip)

File: lib/secure-boot-lib.robot (hunk 4 above)
Comment (macpijan) [reply lib/secure-boot-lib.robot:167]: @Psotas Please create issue so we can fix that later

**Comment by macpijan** [comment]
@Psotas 

> If this is not there, the next tests will skip on Download ISO And Mount As USB when check on IF "${DUT_CONNECTION_METHOD}" == "pikvm"

This is fixed by:
1. Restoring initial method when rebooting etc.
2. Restoring it in `Test Setup`.

**Comment by macpijan** [comment]
1. Results on MSI DDR5:

```
 robot -L TRACE  -v rte_ip:192.168.10.188 -v config:msi-pro-z690-a-ddr5 -v snipeit:no -v device_ip:192.168.10.93 dasharo-security/secure-boot.robot

==============================================================================
Secure-Boot                                                                   
==============================================================================
SBO001.001 Check Secure Boot default state (firmware) :: This test... | PASS |
------------------------------------------------------------------------------
SBO002.001 UEFI Secure Boot (Ubuntu 22.04) :: This test verifies t... | PASS |
------------------------------------------------------------------------------
SBO002.002 UEFI Secure Boot (Windows 11) :: This test verifies tha... | SKIP |
SBO002.002 not supported
------------------------------------------------------------------------------
SBO003.001 Attempt to boot file with the correct key from Shell (f... | PASS |
------------------------------------------------------------------------------
SBO004.001 Attempt to boot file without the key from Shell (firmwa... | PASS |
------------------------------------------------------------------------------
SBO005.001 Attempt to boot file with the wrong-signed key from She... | PASS |
------------------------------------------------------------------------------
SBO006.001 Reset Secure Boot Keys option availability (firmware) :... | PASS |
------------------------------------------------------------------------------
SBO007.001 Attempt to boot the file after restoring keys to defaul... | PASS |
------------------------------------------------------------------------------
SBO008.001 Attempt to enroll the key in the incorrect format (firm... | PASS |
------------------------------------------------------------------------------
Secure-Boot                                                           | PASS |
9 tests, 8 passed, 0 failed, 1 skipped
==============================================================================
```

2. Results on QEMU

- used latest OVMF from: https://github.com/Dasharo/edk2
- used script to start QEMU from: https://github.com/Dasharo/open-source-firmware-validation/pull/116

```
robot  -L TRACE -v config:qemu -v rte_ip:127.0.0.1 -v snipeit:no dasharo-security/secure-boot.robot 
==============================================================================
Secure-Boot                                                                   
==============================================================================
SBO001.001 Check Secure Boot default state (firmware) :: This test... | PASS |
------------------------------------------------------------------------------
SBO002.001 UEFI Secure Boot (Ubuntu 22.04) :: This test verifies t... | PASS |
------------------------------------------------------------------------------
SBO002.002 UEFI Secure Boot (Windows 11) :: This test verifies tha... | SKIP |
SBO002.002 not supported
------------------------------------------------------------------------------
SBO003.001 Attempt to boot file with the correct key from Shell (f... | PASS |
------------------------------------------------------------------------------
SBO004.001 Attempt to boot file without the key from Shell (firmwa... | PASS |
------------------------------------------------------------------------------
SBO005.001 Attempt to boot file with the wrong-signed key from She... | PASS |
------------------------------------------------------------------------------
SBO006.001 Reset Secure Boot Keys option availability (firmware) :... | PASS |
------------------------------------------------------------------------------
SBO007.001 Attempt to boot the file after restoring keys to defaul... | PASS |
------------------------------------------------------------------------------
SBO008.001 Attempt to enroll the key in the incorrect format (firm... | PASS |
------------------------------------------------------------------------------
Secure-Boot                                                           | PASS |
9 tests, 8 passed, 0 failed, 1 skipped
==============================================================================
```

File: lib/secure-boot-lib.robot (hunk 4 above)
Comment (Psotas) [reply lib/secure-boot-lib.robot:167]: @macpijan Issue added https://github.com/Dasharo/open-source-firmware-validation/issues/135
